### PR TITLE
Move to a consistent debug/assert pattern in device_interface.

### PR DESCRIPTION
**Pull request by zvookin**

### CODE REVIEW

File: src/runtime/device_interface.cpp [hunk 1]
```diff
@@ -54,6 +54,20 @@ WEAK int copy_to_host_already_locked(void *user_context, struct halide_buffer_t
...
+namespace {
+
+__attribute__((always_inline))
+void debug_log_and_validate_buf(void * user_context, const halide_buffer_t &buf,
+                                const char *routine) {
+    debug(user_context) << routine << " validating input buffer: " << buf << "\n";
+
+    bool device_interface_set = (buf.device_interface != NULL);
+    bool device_set = (buf.device != 0);
+    halide_assert(user_context, device_interface_set == device_set);
```
Comment (abadams) [inline src/runtime/device_interface.cpp]: A related invariant is (device_set || !buf.device_dirty())

File: src/runtime/device_interface.cpp [hunk 2]
```diff
@@ -54,6 +54,34 @@ WEAK int copy_to_host_already_locked(void *user_context, struct halide_buffer_t
 
 }}} // namespace Halide::Runtime::Internal
 
+namespace {
+
+__attribute__((always_inline))
+void debug_log_and_validate_buf(void * user_context, const halide_buffer_t &buf,
```
Comment (abadams) [inline src/runtime/device_interface.cpp]: I think this should return an error code, in case someone has overridden halide_error to not abort

**Comment by ronlieb** [comment]
This patch seems to have introduced an assert/error in HelloHexagon, when running Offload to HVX

 

./bin/process-host hvx64 1

Error: Buffer has both host and device dirty bits set.

Using HVX 64 schedule

Running pipeline...

Hexagon simulator executed function 0x640e0 in 3691000 cycles

Hexagon simulator executed function 0x640e0 in 3691000 cycles

Hexagon simulator executed function 0x640e0 in 3691000 cycles

Hexagon simulator executed function 0x640e0 in 3691000 cycles

Hexagon simulator executed function 0x640e0 in 3691000 cycles

Hexagon simulator executed function 0x640e0 in 3691000 cycles

Hexagon simulator executed function 0x640e0 in 3691000 cycles

Hexagon simulator executed function 0x640e0 in 3691000 cycles

Hexagon simulator executed function 0x640e0 in 3691000 cycles

Hexagon simulator executed function 0x640e0 in 3691000 cycles

Done, time: 1.37292 s

Success!

 

---

Qualcomm Innovation Center, Inc. is a member of Code Aurora Forum, hosted by The Linux Foundation

**Comment by zvookin** [comment]
I'll take a look at it. Having both dirty bits set is almost always a sign of an error worth fixing.

**Pull request merged**